clients/clients-integration-tests/src/test/java/org/apache/kafka/clients/producer/ProducerFailureHandlingTest.java

```diff
@@ -0,0 +1,305 @@
+            clusterInstance.waitForTopic(Topic.GROUP_METADATA_TOPIC_NAME, 0);
+        }
+
+        try (Producer<byte[], byte[]> producer = clusterInstance.producer(producerConfig(1))) {
+            Exception thrown = assertThrows(ExecutionException.class,
+                    () -> producer.send(new ProducerRecord<>(Topic.GROUP_METADATA_TOPIC_NAME, "test".getBytes(),
+                            "test".getBytes())).get());
+            assertInstanceOf(InvalidTopicException.class, thrown.getCause(),
+                    () -> "Unexpected exception while sending to an invalid topic " + thrown.getCause());
+        }
+    }
+
+    @ClusterTest
+    void testNotEnoughReplicasAfterBrokerShutdown(ClusterInstance clusterInstance) throws InterruptedException,
+            ExecutionException {
+        String topicName = "minisrtest2";
+        int brokerNum = clusterInstance.brokers().size();
+        Map<String, String> topicConfig = Map.of(MIN_IN_SYNC_REPLICAS_CONFIG, String.valueOf(brokerNum));
+        try (Admin admin = clusterInstance.admin()) {
+            admin.createTopics(List.of(new NewTopic(topicName, 1, (short) brokerNum).configs(topicConfig)));
+        }
+
+        ProducerRecord<byte[], byte[]> record =
+                new ProducerRecord<>(topicName, null, "key".getBytes(), "value".getBytes());
+
+        try (Producer<byte[], byte[]> producer = clusterInstance.producer(producerConfig(-1))) {
+            // this should work with all brokers up and running
+            producer.send(record).get();
+            // shut down one broker
+            KafkaBroker oneBroker = clusterInstance.brokers().get(0);
+            oneBroker.shutdown();
+            oneBroker.awaitShutdown();
+
+            Exception e = assertThrows(ExecutionException.class, () -> producer.send(record).get());
+            assertTrue(e.getCause() instanceof NotEnoughReplicasException ||
+                    e.getCause() instanceof NotEnoughReplicasAfterAppendException ||
+                    e.getCause() instanceof TimeoutException);
+
+            // restart the server
+            oneBroker.startup();
+        }
+    }
+
+    private void checkTooLargeRecordForReplicationWithAckAll(ClusterInstance clusterInstance, int maxFetchSize) throws InterruptedException, ExecutionException {
+        int maxMessageSize = maxFetchSize + 100;
+        int brokerSize = clusterInstance.brokers().size();
+        Map<String, String> topicConfig = new HashMap<>();
+        topicConfig.put(MIN_IN_SYNC_REPLICAS_CONFIG, String.valueOf(brokerSize));
+        topicConfig.put(TopicConfig.MAX_MESSAGE_BYTES_CONFIG, String.valueOf(maxMessageSize));
+
+        // create topic
+        String topic10 = "topic10";
+        try (Admin admin = clusterInstance.admin()) {
+            admin.createTopics(List.of(new NewTopic(topic10, brokerSize, (short) brokerSize).configs(topicConfig)));
+            clusterInstance.waitTopicDeletion("topic10");
```

Comment:
Sorry for chiming in late. It's not clear to me why we need to wait for topic deletion here. We only create the topic,  but never delete it.